### PR TITLE
[BugFix] Remove workgroup_id and version pair from workgroup_versions if the workgroup is removed

### PR DESCRIPTION
**Pull request by satanson**
## Why I'm doing:

https://github.com/StarRocks/StarRocksTest/issues/6724

The workgroup is nullptr,  BE crashes inside starrocks::workgroup::WorkGroup::use_big_query_mem_limit() const 

```
(gdb) thread 6
[Switching to thread 6 (Thread 0x7fba31d79700 (LWP 50613))]
#0  starrocks::workgroup::WorkGroup::use_big_query_mem_limit (this=0x0) at /root/starrocks/be/src/exec/workgroup/work_group.h:199
199	        return 0 < _big_query_mem_limit && _big_query_mem_limit <= _mem_tracker->limit();
(gdb) p this
$1 = (const starrocks::workgroup::WorkGroup * const) 0x0
```

When create the a workgroup, the write lock is released and acquired again in the middle of create_workgroup_unlocked's calling add_metrics_unlocked, during this period, the workgroup is deleted but the 
workgroup id and version pair is not deleted from workgroup_versoins. when create_workgroup_unlocked is returned, fragment_executor find that workgroup id and version pair exists in workgroup_versoins, so it directly use operator[] to obtain the workgroup, which leads that a default nullptr object is inserted since the workgroup does not exists.

## What I'm doing:
Use unordered_map::at instead of operator[], and check element existence strictly.
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Review by LiShuMing — DISMISSED** [review]

File: be/src/exec/workgroup/work_group.cpp [hunk 1]
```diff
@@ -272,8 +273,9 @@ WorkGroupPtr WorkGroupManager::add_workgroup(const WorkGroupPtr& wg) {
     std::unique_lock write_lock(_mutex);
     auto unique_id = wg->unique_id();
     create_workgroup_unlocked(wg, write_lock);
```
Comment (trueeyu) [inline be/src/exec/workgroup/work_group.cpp:275]: Why unlock in advance？

File: be/src/exec/workgroup/work_group.cpp [hunk 2]
```diff
@@ -272,8 +273,9 @@ WorkGroupPtr WorkGroupManager::add_workgroup(const WorkGroupPtr& wg) {
...
-    if (_workgroup_versions.count(wg->id()) && _workgroup_versions[wg->id()] == wg->version()) {
-        return _workgroups[unique_id];
+    if (_workgroup_versions.count(wg->id()) && _workgroup_versions[wg->id()] == wg->version() &&
+        _workgroups.count(unique_id)) {
```
Comment (trueeyu) [inline be/src/exec/workgroup/work_group.cpp:277]: Will there be any concurrency issues?

Comment (satanson) [reply be/src/exec/workgroup/work_group.cpp:277]: it is protected by the write lock

**Review by ZiheLiu — DISMISSED** [review]

File: be/src/exec/workgroup/work_group.cpp [hunk 3]
```diff
@@ -482,11 +484,15 @@ void WorkGroupManager::apply(const std::vector<TWorkGroupOp>& ops) {
     while (it != _workgroup_expired_versions.end()) {
         auto wg_it = _workgroups.find(*it);
         if (wg_it != _workgroups.end() && wg_it->second->is_removable()) {
-            int128_t wg_id = *it;
+            auto id = wg_it->second->id();
+            auto version = wg_it->second->version();
             _sum_cpu_limit -= wg_it->second->cpu_limit();
             _workgroups.erase(wg_it);
+            if (_workgroup_versions.count(id) && _workgroup_versions.at(id) <= version) {
```
Comment (imay) [inline be/src/exec/workgroup/work_group.cpp:491]: the find execute twice.
Why not use find() to get an iterator and erase the iterator too.

Comment (satanson) [reply be/src/exec/workgroup/work_group.cpp:491]: It is a extremely low-frequent operations.

**Review by ZiheLiu — DISMISSED** [review]

File: be/src/exec/workgroup/work_group.cpp [hunk 4]
```diff
@@ -273,10 +274,11 @@ WorkGroupPtr WorkGroupManager::add_workgroup(const WorkGroupPtr& wg) {
     auto unique_id = wg->unique_id();
     create_workgroup_unlocked(wg, write_lock);
     if (_workgroup_versions.count(wg->id()) && _workgroup_versions[wg->id()] == wg->version()) {
-        return _workgroups[unique_id];
-    } else {
-        return get_default_workgroup_unlocked();
+        if (auto workgroup_it = _workgroups.find(unique_id); workgroup_it != _workgroups.end()) {
```
Comment (imay) [inline be/src/exec/workgroup/work_group.cpp:277]:
```suggestion
        auto workgroup_it = _workgroups.find(unique_id); 
        if (workgroup_it != _workgroups.end()) {
```

File: be/src/exec/workgroup/work_group.cpp [hunk 5]
```diff
@@ -19,6 +19,7 @@
 #include "common/config.h"
 #include "exec/workgroup/work_group_fwd.h"
 #include "glog/logging.h"
+#include "gutil/strings/substitute.h"
```
Comment (imay) [inline be/src/exec/workgroup/work_group.cpp:22]: seems no need to include this header?

**Review by trueeyu — DISMISSED** [review]

**Review by trueeyu — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by ZiheLiu** [comment]
`add_workgroup(new_wg)`
1. lock.
2. `create_workgroup_unlocked` writes `_workgroups[<wg_id,wg_version>]=new_wg`.
3. `create_workgroup_unlocked` writes `_workgroup_versions[wg_id]=wg_version`.
4. `add_metrics_unlocked` unlocks, does some metrics logic, and relocks.
    - In the meanwhile, the other thread deletes this workgroup by `_workgroups.erase(<wg_id,wg_version>)`, but not erase from `_workgroup_versions`.
5. if `_workgroup_versions[wg_id] == new_wg.version`, then return `_workgroups[<new_wg.id, new_wg.version>]`, which is a nullptr.

**Comment by ZiheLiu** [comment]
triger condition
- After creating a resource, deleting it soon.
- Because `ALTER` is delete and create a new one, so two `ALTER` may also triger this problem.

**Pull request merged**